timsieved/actions.c

```diff
@@ -194,218 +215,182 @@ int capabilities(struct protstream *conn, sasl_conn_t *saslconn,
-  sieve_script_t *s = NULL;
-
-  if (!sievedir_valid_name(name))
-  {
-      prot_printf(conn,"NO \"Invalid script name\"\r\n");
-      return TIMSIEVE_FAIL;
-  }
-
-  if (verify_only) {
-      result = sieve_script_parse_string(interp, buf_cstring(data), &err, &s);
-      sieve_script_free(&s);
-      if (result != SIEVE_OK) result = SIEVEDIR_INVALID;
-  }
-  else {
-      /* see if this would put the user over quota */
-      maxscripts = config_getint(IMAPOPT_SIEVE_MAXSCRIPTS);
-
-      if (sievedir_num_scripts(sieve_dir, name->s)+1 > maxscripts)
-      {
-          prot_printf(conn,
-                      "NO (QUOTA/MAXSCRIPTS) \"You are only allowed %d scripts on this server\"\r\n",
-                      maxscripts);
-          return TIMSIEVE_FAIL;
-      }
-
-      result = sievedir_put_script(sieve_dir,
-                                   buf_cstring(name), buf_cstring(data), &err);
-      if (result == SIEVEDIR_OK)
-          sync_log_sieve(sieved_userid);
-  }
-
-  switch (result) {
-  case SIEVEDIR_INVALID:
-      if (err) {
-          prot_printf(conn, "NO ");
-          prot_printstring(conn, err);
-          prot_printf(conn, "\r\n");
-          free(err);
-      } else {
-          prot_printf(conn, "NO \"parse failed\"\r\n");
-      }
-      return TIMSIEVE_FAIL;
-
-    case SIEVEDIR_FAIL:
-        prot_printf(conn, "NO \"bytecode generate failed\"\r\n");
-        return TIMSIEVE_FAIL;
+    int result;
+    char *err = NULL;
+    int numscripts, maxscripts;
+
+    if (!verify_only) {
+        if (!sievedir_valid_name(name)) {
+            prot_printf(conn,"NO \"Invalid script name\"\r\n");
+            return TIMSIEVE_FAIL;
+        }
 
-    case SIEVEDIR_IOERROR:
-        prot_printf(conn, "NO \"%s\"\r\n", strerror(errno));
+        /* see if this would put the user over quota */
+        maxscripts = config_getint(IMAPOPT_SIEVE_MAXSCRIPTS);
+
+        sievedb_count(sievedb, &numscripts);
+        if (numscripts == maxscripts) {
+            prot_printf(conn, "NO (QUOTA/MAXSCRIPTS)"
+                        " \"You are only allowed %d scripts on this server\"\r\n",
+                        maxscripts);
+            return TIMSIEVE_FAIL;
+        }
+    }
+
+    result = sieve_script_parse_string(interp, buf_cstring(data), &err, NULL);
+
+    if (result != SIEVE_OK) {
+        if (err) {
+            prot_printf(conn, "NO ");
+            prot_printstring(conn, err);
+            prot_printf(conn, "\r\n");
+            free(err);
+        } else {
+            prot_printf(conn, "NO \"parse failed\"\r\n");
+        }
         return TIMSIEVE_FAIL;
     }
 
-  prot_printf(conn, "OK\r\n");
+    if (!verify_only) {
+        struct sieve_data *sdata = NULL;
 
-  return TIMSIEVE_OK;
-}
+        result = sievedb_lookup_name(sievedb, mailbox_name(sieve_mailbox),
+                                     buf_cstring(name), &sdata, 0);
+        if (!result || result == CYRUSDB_NOTFOUND) {
+            sdata->name = buf_cstring(name);
```

Comment:
I don't love this design where there's a `static struct sieve_data` inside `sievedb_lookup_name` when gets filled and zeroed regardless of whether data is found or not.  Should we be defining the struct out here and passing it into the lookup instead?

Reply:
This is the same as caldav_db and carddav_db.  Should we change all of them?

Reply:
Since the same issue is present in caldav_db, carddav_cb, and webdav_db, will wait to resolve this until after this PR lands on master.